### PR TITLE
SSE/Alerting: Support prom instant vector responses

### LINKED ISSUE #35663: Alerting: Support Prometheus Vector (Number/SingleStat) responses with Grafana Managed Alerts
**What would you like to be added**:
When using the prometheus data source from a "Grafana managed alert rule", a Reduce or Classic Condition (Server Side Expression) SSE is needed currently. If we support instant-vector queries on the backend prom queries could be directly alerted on with a Grafana managed alert.

This is because the current _backend_ data source implementation doesn't support instant-vector responses (so we do not get reduced data back from prometheus):

https://github.com/grafana/grafana/blob/61a718cf8d9feb7c2ac8fde5cb65ff4a383d6c20/pkg/tsdb/prometheus/prometheus.go#L167-L174

Note: Would also need to make sure the instant-vector selector ends up in the backend model.

Additionally on the backend, we don't have a clear defined shape of dataframe(s) for a labeled numeric response, so we should pick that first to get more consistency across data sources.

Perhaps we can base this on whatever shape of frame the current frontend uses.

We currently support direct alerting of number for Table like data sources, e.g. SQL: https://grafana.com/docs/grafana/latest/alerting/unified-alerting/grafana-managed-numeric-rule/ .

**Why is this needed**:
Follow up to https://github.com/grafana/grafana/issues/35540 . Since once can do prom ruler managed alerts with the new unified alerting system I think that lowers the priority of this a little. But once we pick the type I don't think it will be hard to do.

### PR DESCRIPTION
**Pull request by kylebrandt**
**What this PR does / why we need it**:
Recognizes an instant-vector response from Prom, and maps to an SSE NumberSet, so user can do things like alert directly on it within a Grafana Managed alert.

image

Also, it updates reduce and classic expressions to support set of `number` instead of just `series`. This is needed to keep the change backward compatible: i.e. keep the same behavior if a user applies a reduce command or a classic expression to the result of the query.
- Reduce expression just ignores number set and returns it as is, without any changes. Mathematically, this is the correct behavior because any reduce operation applied to a set with a single value - results in the same value. Also, it sets notice in the value's metadata 
<details> <summary>Example Response with Notice </summary> 

```json
{
  "results": {
    "A": {
      "frames": [
        {
          "schema": {
            "refId": "A",
            "fields": [
              {
                "name": "up{instance=\"fake-prometheus-data:9091\", job=\"fake-data-gen\"}",
                "type": "number",
                "typeInfo": {
                  "frame": "float64",
                  "nullable": true
                },
                "labels": {
                  "__name__": "up",
                  "instance": "fake-prometheus-data:9091",
                  "job": "fake-data-gen"
                }
              }
            ]
          },
          "data": {
            "values": [
              [
                1
              ]
            ]
          }
        },
        {
          "schema": {
            "refId": "A",
            "fields": [
              {
                "name": "up{instance=\"host.docker.internal:3000\", job=\"grafana\"}",
                "type": "number",
                "typeInfo": {
                  "frame": "float64",
                  "nullable": true
                },
                "labels": {
                  "__name__": "up",
                  "instance": "host.docker.internal:3000",
                  "job": "grafana"
                }
              }
            ]
          },
          "data": {
            "values": [
              [
                1
              ]
            ]
          }
        },
        {
          "schema": {
            "refId": "A",
            "fields": [
              {
                "name": "up{instance=\"localhost:9090\", job=\"prometheus\"}",
                "type": "number",
                "typeInfo": {
                  "frame": "float64",
                  "nullable": true
                },
                "labels": {
                  "__name__": "up",
                  "instance": "localhost:9090",
                  "job": "prometheus"
                }
              }
            ]
          },
          "data": {
            "values": [
              [
                1
              ]
            ]
          }
        },
        {
          "schema": {
            "refId": "A",
            "fields": [
              {
                "name": "up{instance=\"node_exporter:9100\", job=\"node_exporter\"}",
                "type": "number",
                "typeInfo": {
                  "frame": "float64",
                  "nullable": true
                },
                "labels": {
                  "__name__": "up",
                  "instance": "node_exporter:9100",
                  "job": "node_exporter"
                }
              }
            ]
          },
          "data": {
            "values": [
              [
                1
              ]
            ]
          }
        },
        {
          "schema": {
            "refId": "A",
            "fields": [
              {
                "name": "up{instance=\"prometheus-random-data:8080\", job=\"prometheus-random-data\"}",
                "type": "number",
                "typeInfo": {
                  "frame": "float64",
                  "nullable": true
                },
                "labels": {
                  "__name__": "up",
                  "instance": "prometheus-random-data:8080",
                  "job": "prometheus-random-data"
                }
              }
            ]
          },
          "data": {
            "values": [
              [
                1
              ]
            ]
          }
        }
      ]
    },
    "B": {
      "frames": [
        {
          "schema": {
            "refId": "B",
            "meta": {
              "notices": [
                {
                  "severity": "warning",
                  "text": "Query or expression A produced result of type numberSet. Reduce operation is not necessary."
                }
              ]
            },
            "fields": [
              {
                "name": "B",
                "type": "number",
                "typeInfo": {
                  "frame": "float64",
                  "nullable": true
                },
                "labels": {
                  "__name__": "up",
                  "instance": "fake-prometheus-data:9091",
                  "job": "fake-data-gen"
                }
              }
            ]
          },
          "data": {
            "values": [
              [
                1
              ]
            ]
          }
        },
        {
          "schema": {
            "refId": "B",
            "meta": {
              "notices": [
                {
                  "severity": "warning",
                  "text": "Query or expression A produced result of type numberSet. Reduce operation is not necessary."
                }
              ]
            },
            "fields": [
              {
                "name": "B",
                "type": "number",
                "typeInfo": {
                  "frame": "float64",
                  "nullable": true
                },
                "labels": {
                  "__name__": "up",
                  "instance": "host.docker.internal:3000",
                  "job": "grafana"
                }
              }
            ]
          },
          "data": {
            "values": [
              [
                1
              ]
            ]
          }
        },
        {
          "schema": {
            "refId": "B",
            "meta": {
              "notices": [
                {
                  "severity": "warning",
                  "text": "Query or expression A produced result of type numberSet. Reduce operation is not necessary."
                }
              ]
            },
            "fields": [
              {
                "name": "B",
                "type": "number",
                "typeInfo": {
                  "frame": "float64",
                  "nullable": true
                },
                "labels": {
                  "__name__": "up",
                  "instance": "localhost:9090",
                  "job": "prometheus"
                }
              }
            ]
          },
          "data": {
            "values": [
              [
                1
              ]
            ]
          }
        },
        {
          "schema": {
            "refId": "B",
            "meta": {
              "notices": [
                {
                  "severity": "warning",
                  "text": "Query or expression A produced result of type numberSet. Reduce operation is not necessary."
                }
              ]
            },
            "fields": [
              {
                "name": "B",
                "type": "number",
                "typeInfo": {
                  "frame": "float64",
                  "nullable": true
                },
                "labels": {
                  "__name__": "up",
                  "instance": "node_exporter:9100",
                  "job": "node_exporter"
                }
              }
            ]
          },
          "data": {
            "values": [
              [
                1
              ]
            ]
          }
        },
        {
          "schema": {
            "refId": "B",
            "meta": {
              "notices": [
                {
                  "severity": "warning",
                  "text": "Query or expression A produced result of type numberSet. Reduce operation is not necessary."
                }
              ]
            },
            "fields": [
              {
                "name": "B",
                "type": "number",
                "typeInfo": {
                  "frame": "float64",
                  "nullable": true
                },
                "labels": {
                  "__name__": "up",
                  "instance": "prometheus-random-data:8080",
                  "job": "prometheus-random-data"
                }
              }
            ]
          },
          "data": {
            "values": [
              [
                1
              ]
            ]
          }
        }
      ]
    }
  }
}
```
</details>
- Classic condition performs a reduction of the numbers the same way it does with series.

**Which issue(s) this PR fixes**:

Fixes #35663

**Special notes for your reviewer**:

Very quick (linear code writing), not much manual testing and no actual tests. So need someone on alerting to perhaps pick this up and make sure it doesn't mess up anything - was looking to bootstrap for someone. 

Some things to check:
 - [ ] Zero length Field responses to handle? (see code comment)
 - [ ] Does metric name survive if it is pointed to directly in the condition, may become RefID ... can't recall. Perhaps doesn't matter much
 - [ ] Does this break anything with series?

### CODE REVIEW

File: pkg/expr/nodes.go [hunk 1]
```diff
@@ -236,6 +236,41 @@ func (dn *DSNode) Execute(ctx context.Context, vars mathexp.Vars, s *Service) (m
 			return mathexp.Results{}, QueryError{RefID: refID, Err: qr.Error}
 		}
 
+		dataSource := dn.datasource.Type
+		if dataSource == "prometheus" {
```
Comment (kylebrandt) [inline pkg/expr/nodes.go:240]: Maybe make this and the block below it extracted into funcs

Comment (torkelo) [reply pkg/expr/nodes.go:240]: What is special about a vector result that needs special handling?

Comment (kylebrandt) [reply pkg/expr/nodes.go:240]: SSE has similar types to Prometheus, Matrix/Vector (Except TimeSeries / Number in the case of SSE). The number type can be alerted on directly, whereas series need to be reduced. So without this indicator, SSE wouldn't know if a response is a time series of length 1, or a number response.

**Comment by kylebrandt** [comment]
> Reduce expression just ignores number set and returns it as is, without any changes. Mathematically, this is the correct behavior because any reduce operation applied to a set with a single value - results in the same value.

^ This is what concerns me the most, but not sure of other ways forward... My concern is that if a user is reducing a numberSet, in any other case but this but this it would be a user mistake -- as in they probably selected the wrong RefID as an input.

Short term:
 - Can we flag the response when SSE reads it, and only allow it if the data source is prometheus?

Longer Term:
 - WG fro data plane contract is actually starting to move forward, so eventually these response will have type/format/version (or something along these lines), so might be able to swtich behavior on that (old vs new, deprecate old over time). Although more proper would be to have query versions as well, as that info would correspond to creation time vs runtime, and allow us to link behavior directly.

So by doing it only with prometheus, can hopefully keep the design intact more clearly until the versioned behavior in place.

File: pkg/expr/transform.go [hunk 2]
```diff
@@ -56,7 +56,7 @@ type TimeRange struct {
 
 // TransformData takes Queries which are either expressions nodes
 // or are datasource requests.
-func (s *Service) TransformData(ctx context.Context, req *Request) (r *backend.QueryDataResponse, err error) {
+func (s *Service) TransformData(ctx context.Context, req *Request) (r *Response, err error) {
```
Comment (kylebrandt) [inline pkg/expr/transform.go:59]: I believe we should keep these the same type so there are not blockers to extracting the service in the future.

So I would rather see one of the two ugly choices:
 - A) Attach it the the first frame (preferred)
 - B) Attach it to all frames

And then look for it somewhere outside the service (like the FE if that is possible I think would be best)

This also means when there is no data, the warning will be omitted, which is also unfortunate but okay I think. In someday-future, I have a design where an empty response always include at least one frame (with length 0/nil) to indicate an empty response.  So once that contract is in place that will be solvable. (Or, as long as alerting understands it correctly as NoData, SSE could start to implement this now on output. There may be some FE concerns outside of alerting here but I don't think this is normalized yet anyways.)

Comment (yuri-tceretian) [reply pkg/expr/transform.go:59]: As an option, we go with option B) and then ask FE to dedup messages. In that case if there is the same notice in all frames, FE will squash it and show just a single warning. In the case of different notices, FE will show them all.

Comment (kylebrandt) [reply pkg/expr/transform.go:59]: That works as well as long as we keep the information needed in `[]data.Frame`, (well `[]*data.Frame`, but that is a mistake that would be nice to get to one day as far as I can tell).

File: pkg/expr/commands.go [hunk 3]
```diff
@@ -152,18 +153,27 @@ func (gr *ReduceCommand) NeedsVars() []string {
...
+			if err != nil {
+				return newRes, err
+			}
+			newRes.Values = append(newRes.Values, num)
+		case mathexp.Number: // if incoming vars is just a number, any reduce op is just a noop, add it as it is
+			copyV := mathexp.NewNumber(gr.refID, v.GetLabels())
+			copyV.SetValue(v.GetFloat64Value())
+			copyV.AddNotice(data.Notice{
+				Severity: data.NoticeSeverityWarning,
+				Text:     fmt.Sprintf("Query or expression %s produced result of type 'single number'. Reduce operation is not needed.", gr.VarToReduce),
```
Comment (kylebrandt) [inline pkg/expr/commands.go:171]: Maybe "Reduce operation is not needed. Input query or expression %s is already reduced data."

"

**Review by yuri-tceretian — APPROVED** [review]
LGTM

**Comment by yuri-tceretian** [comment]
@grafana/observability-metrics and\or @grafana/backend-platform  can some of you look at it?

**Pull request merged**